### PR TITLE
Sb 1420

### PR DESCRIPTION
**Pull request by amit-tarento**

### CODE REVIEW

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProviderFactory.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.keycloak.services.resource.RealmResourceProviderFactory;
+
+/**
+ * 
+ * @author Amit Kumar
+ *
+ */
+public class RestResourceProviderFactory implements RealmResourceProviderFactory {
+
+  public static final String ID = "getlink";
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProviderFactory.java]: get-required-action-link

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProviderFactory.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Override
+  public String getId() {
+    return ID;
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProviderFactory.java]: Add debug logger statements.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProviderFactory.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.services.resource.RealmResourceProvider;
+import org.keycloak.services.resource.RealmResourceProviderFactory;
+
+/**
+ * 
+ * @author Amit Kumar
+ *
+ */
+public class RestResourceProviderFactory implements RealmResourceProviderFactory {
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProviderFactory.java]: RequiredActionLinkProviderFactory.java
RequiredActionLinkProvider.java

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+/**
+ * 
+ * @author Amit Kumar
+ *
+ */
+public class RestResourceProvider implements RealmResourceProvider {
+
+  private KeycloakSession session;
+  private AuthResult auth;
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: auth -> authResult

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+   * create a link on which the user can click to reset their password or verify email based on
+   * action param value.
+   *
+   * @param request Map<String,String> request object which contains details of redirectUri,client-id,action,userName
+   * @return
+   */
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response generateLink(Map<String, String> request) {
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: generateRequiredActionLink

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    String redirectUri = request.get(Constants.REDIRECT_URI);
+    String clientId = request.get(Constants.CLIENT_ID);
+    String action = request.get(Constants.ACTION);
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: ACTION -> REQUIRED_ACTION

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    String userName = request.get(Constants.USERNAME);
+    Integer lifespan = null;
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: lifespan -> expirationInSecs
LIFE_SPAN -> EXPIRATION_IN_SECS

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/utils/Constants.java [hunk 8]
```diff
@@ -7,5 +7,27 @@ private Constants(){}
   public static final String MULTIPLE_USER_ASSOCIATED_WITH_PHONE = "Multiple users are associated with this phone.";
   public static final String MULTIPLE_USER_ASSOCIATED_WITH_EMAIL = "Multiple users are associated with this email.";
   public static final String MULTIPLE_USER_ASSOCIATED_WITH_USERNAME = "Multiple users are associated with this username.";
+  public static final String REDIRECT_URI = "redirectUri";
+  public static final String CLIENT_ID = "clientId";
+  public static final String ACTION = "action";
+  public static final String USERNAME = "userName";
+  public static final String LIFE_SPAN = "lifespan";
+  public static final String INVALID_LIFESPAN_VALUE = "Invalid lifespan value: ";
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/utils/Constants.java]: Invalid value {0} for parameter {1}.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+          checkRealmAdmin();
+        } catch (NotAuthorizedException ex) {
+          return ErrorResponse.error(Constants.NOT_AUTHORIZED, Status.UNAUTHORIZED);
+        } catch (ForbiddenException ex) {
+          return ErrorResponse.error(Constants.DOES_NOT_HAVE_REALM_ADMIN_ROLE, Status.FORBIDDEN);
+        }
+      }
+    }
+    List<String> actions = new LinkedList<>();
+    if (UserModel.RequiredAction.UPDATE_PASSWORD.name().equalsIgnoreCase(action)) {
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: Use switch statement.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+          // Validate the Authorization key in header
+          checkRealmAdmin();
+        } catch (NotAuthorizedException ex) {
+          return ErrorResponse.error(Constants.NOT_AUTHORIZED, Status.UNAUTHORIZED);
+        } catch (ForbiddenException ex) {
+          return ErrorResponse.error(Constants.DOES_NOT_HAVE_REALM_ADMIN_ROLE, Status.FORBIDDEN);
+        }
+      }
+    }
+    List<String> actions = new LinkedList<>();
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: actions -> actionName

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+      redirect = RedirectUtils.verifyRedirectUri(session.getContext().getUri(), redirectUri,
+          session.getContext().getRealm(), client);
+      if (redirect == null) {
+        throw new WebApplicationException(
+            ErrorResponse.error(Constants.INVALID_REDIRECT_URI, Status.BAD_REQUEST));
+      }
+    }
+
+    if (lifespan == null) {
+      lifespan = 7200;
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: Define constant: DEFAULT_LINK_EXPIRATION_IN_SECS

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    if (user == null) {
+      return ErrorResponse.error(Constants.INVALID_USERNAME + userName, Status.BAD_REQUEST);
+    }
+
+    if (!user.isEnabled()) {
+      throw new WebApplicationException(
+          ErrorResponse.error(Constants.USER_IS_DISABLED, Status.BAD_REQUEST));
+    }
+
+    if (redirectUri != null && clientId == null) {
```
Comment (bvinayakumar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RestResourceProvider.java]: Validate for blank redirect uri.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+      
+      Map<String, Object> response = new HashMap<>();
+      response.put(Constants.LINK, link);
+      return Response.ok(response).build();
+    } catch (Exception e) {
+      return ErrorResponse.error(Constants.ERROR_CREATE_LINK, Status.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  private UserModel getUserByUserName(String userName) {
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java]: suggest rename to `getEnabledUserByUsernameOrError`

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+    if (!user.isEnabled()) {
+      throw new WebApplicationException(
+          ErrorResponse.error(Constants.ERROR_USER_IS_DISABLED, Status.BAD_REQUEST));
+    }
+    
+    return user;
+  }
+
+  private List<String> getRequiredActionList(String actionName) {
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java]: Suggest similar rename as above indicating that calling this method may cause the application to return an error.

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+    String redirectUri = request.get(Constants.REDIRECT_URI);
+    String clientId = request.get(Constants.CLIENT_ID);
+    String actionName = request.get(Constants.REQUIRED_ACTION);
+    String userName = request.get(Constants.USERNAME);
+    
+    validateClientId(clientId);
+   
+    UserModel user = getUserByUserName(userName);
+    
+    ClientModel client = session.getContext().getRealm().getClientByClientId(clientId);
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java]: the pattern being used for fetching the client vs fetching the user seem to be different.

For the client the code first calls a method to validate the Id, then fetches the object.
For the user, the code calls a method which validates the id and fetches the object in one call. 

Suggest choosing one of the two approaches and applying it consistently. Possibly update the `validateClientId` method  by renaming it to `getClientByClientIdOrError()`

File: keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+    
+    return requiredActionList;
+  }
+  
+  private int getExpirationInSecs(String expirationInSecsStr) {
+    int expirationInSecs;
+    
+    try {
+      if (StringUtils.isNotBlank(expirationInSecsStr)) {
+        expirationInSecs = Integer.parseInt(expirationInSecsStr);
```
Comment (kochhar) [inline keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java]: Is there a maximum expiration time to enforce?

Comment (amit-tarento) [reply keycloak/sms-provider/src/main/java/org/sunbird/keycloak/rest/RequiredActionLinkProvider.java]: yes max expiration time we can enforce is INT_MAX as the maximum 'int' value

**Comment by bvinayakumar** [comment]
Merged this PR as it is approved by Shailesh.

**Pull request merged**